[Live Range Selection] Expanding selection by granularity should expand anchor & focus

rniwa opened the pull request:
#### ca201a717c242c53d1171e58aee57a5665cd5bc5
<pre>
[Live Range Selection] Expanding selection by granularity should expand anchor &amp; focus
<a href="https://bugs.webkit.org/show_bug.cgi?id=247239">https://bugs.webkit.org/show_bug.cgi?id=247239</a>

Reviewed by Wenson Hsieh and Darin Adler.

This patch reverts commits.webkit.org/r268847 and implements a new fix.

Ideally, we fix VisibleSelection::adjustSelectionRespectingGranularity to not expand the end more than
required but that approach causes regressions (e.g. extending selection backwards will end up
deselecting the original word that was double tapped when drag selecting text backwards), some of which
are hard to overcome.

So, this patch instead detects when the ending is expanding unexpectedly, and corrects it in
EventHandler::updateSelectionForMouseDrag.

* LayoutTests/editing/selection/double-click-expands-focus-anchor-live-range-expected.txt: Added.
* LayoutTests/editing/selection/double-click-expands-focus-anchor-live-range.html: Added.

* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::validate): Reverted r268847.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateSelectionForMouseDownDispatchingSelectStart):
(WebCore::EventHandler::handleMousePressEventDoubleClick):
(WebCore::EventHandler::updateSelectionForMouseDrag): Detect when the end is erroneously extending,
and correct back to the original end.

* Source/WebCore/page/EventHandler.h:

Canonical link: <a href="https://commits.webkit.org/256174@main">https://commits.webkit.org/256174@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/002cc4d652b20ce07be7d57d668b42dd01bb7661

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94937 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4085 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27843 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104542 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164806 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/98932 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4170 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/32268 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87243 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100462 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100606 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3013 "Passed tests") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/81513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29998 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84927 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/84482 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72889 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38677 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18313 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36508 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19594 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4257 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40434 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/81513 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42410 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38828 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebCore/editing/VisibleSelection.cpp:
```diff
@@ -418,8 +418,8 @@ void VisibleSelection::validate(TextGranularity granularity)
 
-    bool shouldUpdateAnchor = false; // Set to false because of <rdar://problem/69542459>. Can be returned to original logic when this problem is fully fixed.
-    bool shouldUpdateFocus = false; // Ditto.
+    bool shouldUpdateAnchor = m_start != startBeforeAdjustments; // Set to false because of <rdar://problem/69542459>. Can be returned to original logic when this problem is fully fixed.
```
> `// Set to false because of <rdar://problem/69542459> …`

Nit - is this comment still relevant?

rniwa replied:
Oh oops, will remove.

whsieh commented on LayoutTests/editing/selection/double-click-expands-focus-anchor-live-range.html:
```diff
@@ -0,0 +1,57 @@
+{
+    getSelection().removeAllRanges();
+    const x = content.offsetLeft + content.offsetWidth / 2;
+    const y = content.offsetTop + content.offsetHeight / 2;
```
Nit - can we just remove these two lines?

rniwa replied:
Done.

whsieh commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1017,9 +1017,19 @@ void EventHandler::updateSelectionForMouseDrag(const HitTestResult& hitTestResul
+
+        if (!newSelection.isBaseFirst() && !oldSelection.isBaseFirst() && oldSelection.end() < newSelection.end())
+            newSelection.setBase(dragStartSelection.end());
+        else if (newSelection.isBaseFirst() && !oldSelection.isBaseFirst() && oldSelection.start() < newSelection.start())
```
Hm...I'm a little bit confused by this check. Is this handling the case where the selection is about to flip? (since the new selection's base is first, and the old selection's base is not first). Or was this intended to be `newSelection.isBaseFirst() && oldSelection.isBaseFirst()`？

rniwa replied:
This is reverting the base back to the start when we're extending selection forwards. If we had adjusted the base in the above code, we need to adjust it back to the start here.

rniwa replied:
To put it another way, the primary fix of this patch does a thing where we re-set "base" to the end of selection when we're extending selection backwards. But when this selection flips back to forward direction, we need to use the selection start as the base instead. That code does that.

rniwa commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -1017,9 +1017,19 @@ void EventHandler::updateSelectionForMouseDrag(const HitTestResult& hitTestResul
 
+        auto dragStartPosition = m_mousePressNode->renderer()->positionForPoint(m_dragStartPosition, nullptr);
+        VisibleSelection dragStartSelection = dragStartPosition;
+        dragStartSelection.expandUsingGranularity(m_frame.selection().granularity());
```
Actually, we don't need this selection. We can just use oldSelection instead.

rniwa replied:
Hm... we'd need to remember the original selection instead.
There is some kind of a bug in positionForPoint...

darinadler commented on Source/WebCore/page/EventHandler.h:
```diff
@@ -606,6 +606,7 @@ class EventHandler {
 
 #if ENABLE(DRAG_SUPPORT)
     LayoutPoint m_dragStartPosition;
+    VisibleSelection m_dragStartSelection;
```
I worry about using `VisibleSelection` for anything that is stored for a length of time. When a document is edited a `VisibleSelection` can become non-canonical, so the class makes promises it can’t keep, claiming that the endpoints are already canonicalized. It seems like if we are storing something for a longer period, we need to store it in some more primitive data structure that can’t have broken invariants. That’s why I was hoping we’d use classes more like SimpleRange to store things. It’s still useful to have the `VisiblePosition` family of classes to make it easier to express converting things to canonical form, but keeping them around after DOM modifications are made always seems like a risk when we use them as data members for a long-lived object or even local variables in a long function.

rniwa replied:
Sure, we can use `std::optional<SimplerRange>` instead.

darinadler reviewed (approved):
If this passes all the tests we have that’s great. I really hope at some point we can move to a simpler less fragile model where more of the work can be done as originally intended in the functions that are supposed to expand to a granularity, but I think it’s important to get to the point where we can turn on the live range semantics soon without requiring us to fix those fundamentals first.

rniwa committed:
[Live Range Selection] Expanding selection by granularity should expand anchor & focus
https://bugs.webkit.org/show_bug.cgi?id=247239

Reviewed by Wenson Hsieh and Darin Adler.

This patch reverts commits.webkit.org/r268847 and implements a new fix.

Ideally, we fix VisibleSelection::adjustSelectionRespectingGranularity to not expand the end more than
required but that approach causes regressions (e.g. extending selection backwards will end up
deselecting the original word that was double tapped when drag selecting text backwards), some of which
are hard to overcome.

So, this patch instead detects when the ending is expanding unexpectedly, and corrects it in
EventHandler::updateSelectionForMouseDrag.

* LayoutTests/editing/selection/double-click-expands-focus-anchor-live-range-expected.txt: Added.
* LayoutTests/editing/selection/double-click-expands-focus-anchor-live-range.html: Added.

* Source/WebCore/editing/VisibleSelection.cpp:
(WebCore::VisibleSelection::validate): Reverted r268847.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateSelectionForMouseDownDispatchingSelectStart):
(WebCore::EventHandler::handleMousePressEventDoubleClick):
(WebCore::EventHandler::updateSelectionForMouseDrag): Detect when the end is erroneously extending,
and correct back to the original end.

* Source/WebCore/page/EventHandler.h:

Canonical link: https://commits.webkit.org/256174@main
